lib/features/composer/presentation/widgets/email_address_input_builder.dart

```diff
@@ -238,27 +259,113 @@ class EmailAddressInputBuilder {
-      subtitle: emailAddress.displayName.isNotEmpty && emailAddress.emailAddress.isNotEmpty
+      subtitle: suggestionEmailAddress.emailAddress.displayName.isNotEmpty && suggestionEmailAddress.emailAddress.emailAddress.isNotEmpty
           ? Text(
-              emailAddress.emailAddress,
+              suggestionEmailAddress.emailAddress.emailAddress,
               maxLines: 1,
               overflow: kIsWeb ? null : TextOverflow.ellipsis,
               style: const TextStyle(color: AppColor.colorHintSearchBar, fontSize: 13, fontWeight: FontWeight.normal))
           : null,
       onTap: () {
-        setState(() => listEmailAddress.add(emailAddress));
+        log('EmailAddressInputBuilder::_buildSuggestionItem(): onTap: $suggestionEmailAddress');
+        setState(() => listEmailAddress.add(suggestionEmailAddress.emailAddress));
         _onUpdateListEmailAddressAction?.call(_prefixEmailAddress, listEmailAddress);
-        tagEditorState.selectSuggestion(emailAddress);
+        tagEditorState.selectSuggestion(suggestionEmailAddress);
       },
     );
   }
 
+  FutureOr<List<SuggestionEmailAddress>> _findSuggestions(String query) async {
+    log('EmailAddressInputBuilder::_buildTagEditor(): findSuggestions: $query');
+    if (_gapBetweenTagChangedAndFindSuggestion?.isActive ?? true) {
+      return [];
+    }
+
+    final processedQuery = query.trim();
+
+    if (processedQuery.isEmpty) {
+      return [];
+    }
+
+    final tmailSuggestion = List<SuggestionEmailAddress>.empty(growable: true);
+    if (processedQuery.isNotEmpty && _onSuggestionEmailAddress != null) {
+      tmailSuggestion.addAll(
+        (await _onSuggestionEmailAddress!(processedQuery))
+          .map((emailAddress) => _toSuggestionEmailAddress(emailAddress, listEmailAddress)));
+    }
+
+    tmailSuggestion.addAll(_matchedSuggestionEmailAddress(processedQuery, listEmailAddress));
+
+    return tmailSuggestion;
+  }
+
   bool _isEmailAddressValid(String emailAddress) => GetUtils.isEmail(emailAddress);
+
+  bool _isDuplicatedRecipient(String inputEmail) {
+    if (inputEmail.isEmpty) {
+      return false;
+    }
+    return listEmailAddress
+      .map((emailAddress) => emailAddress.email)
+      .whereNotNull()
+      .contains(inputEmail);
+  }
+
+  SuggestionEmailAddress _toSuggestionEmailAddress(EmailAddress item, List<EmailAddress> addedEmailAddresses) {
+    if (addedEmailAddresses.contains(item)) {
+      return SuggestionEmailAddress(item, state: SuggestionEmailState.duplicated);
+    } else {
+      return SuggestionEmailAddress(item);
+    }
+  }
+
+  Iterable<SuggestionEmailAddress> _matchedSuggestionEmailAddress(String query, List<EmailAddress> addedEmailAddress) {
+    return addedEmailAddress.where((addedMail) => addedMail.emailAddress.contains(query))
+      .map((emailAddress) => SuggestionEmailAddress(emailAddress, state: SuggestionEmailState.duplicated));
+  }
+
+  void _handleGapBetweenTagChangedAndFindSuggestion() {
```

Comment:
why not handle anything in this function. It's a bit pointless here.

Reply:
because timer need the callback, but I don't handle anything on it.